### PR TITLE
AXLE-1: Create session management service

### PR DESCRIPTION
**Pull request by martaciesielska**
This implementations covers:
1) Axle server - the session management service and its dependent modules)
2) A sample Single Page Application which demonstrates the effects of using Axle and provides sample code in JavaScript (based on the Ironclad SPA)
3) Automated tests (acceptance - black box tests and unit tests)

Couple of considerations:
- this implementation doesn't handle network partitioning (we have a separate story for that)
- this implementation doesn't cover security (we have a separate story for that)
- for simplicity, since network partitioning is not covered, I've temporarily used SignalR connection IDs to distinguish sessions and made the session repository in-memory
- currently, in order to run the example SPA, you need to also run Ironclad with some extra setup - I'll change the settings and perform setup of the test instance so that this is not required
- incoming connections to the SignalR hub are handled on different threads hence the usage of multiple threads in the unit tests covering concurrency issues

### CODE REVIEW

**Comment by martaciesielska** [comment]
Pending:
- conversion of the ugly acceptance test (last commit) into a more sensible unit test
- update the sample SPA to use TEST instance of Ironclad (or add instructions on how to set it up locally... or both)
- fixes for concurrency issues (usage of plain dictionaries)

File: src/Docker/docker-compose.override.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  nginx:
+    image: nginx:1.13.8-alpine
+    depends_on:
+      - api
+    volumes:
+      - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
+      - ${LOG_PATH:-/var/log}:/var/log
+    ports:
+      - "5007:80"
```
Comment (Waqas-Zamir) [inline src/Docker/docker-compose.override.yml]: Please change the port to Axle given port

File: src/Tests/Axle.Tests.Acceptance/NewSessionFeature.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                {
+                    await signalRClient2.StartConnection();
+                    await signalRClient2.StartSession(AxleUserId);
+                });
+
+            "Then the first session gets terminated"
+                .x(async () =>
+                {
+                    await Task.Delay(100);
+                    signalRClient.IsConnected.Should().BeFalse();
```
Comment (Waqas-Zamir) [inline src/Tests/Axle.Tests.Acceptance/NewSessionFeature.cs]: Also confirm if the `signalRClient2.IsConnected` is still true

File: src/Tests/Axle.Tests.Acceptance/testsettings.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿{
+  "axle-url": "http://localhost:5012",
```
Comment (Waqas-Zamir) [inline src/Tests/Axle.Tests.Acceptance/testsettings.json]: nice to use camel casing for property name to align with other property name like `axleUrl` and `clientId`

File: src/Axle/Hubs/SessionHub.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    using Serilog;
+
+    public class SessionHub : Hub
+    {
+        private readonly SessionHubMethods<SessionHub> hubMethods;
+        private readonly IRepository<string, HubConnectionContext> connectionRepository;
+
+        public SessionHub(SessionHubMethods<SessionHub> hubMethods, IRepository<string, HubConnectionContext> connectionRepository)
+        {
+            this.hubMethods = hubMethods ?? throw new ArgumentNullException(nameof(hubMethods));
```
Comment (Waqas-Zamir) [inline src/Axle/Hubs/SessionHub.cs]: I guess DI will perform the argument null check.

File: src/Tests/Axle.Tests.Acceptance/NewSessionFeature.cs (hunk 2 above)
Comment (martaciesielska) [reply src/Tests/Axle.Tests.Acceptance/NewSessionFeature.cs]: Done.

File: src/Docker/docker-compose.dcproj [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="15.0" Sdk="Microsoft.Docker.Sdk">
+  <PropertyGroup Label="Globals">
+    <ProjectVersion>2.0</ProjectVersion>
+    <DockerTargetOS>Linux</DockerTargetOS>
+    <ProjectGuid>E90A2797-21AB-4CB2-AB0C-1C2914545CB1</ProjectGuid>
+    <DockerLaunchBrowser>True</DockerLaunchBrowser>
+    <DockerServiceUrl>http://localhost:5007</DockerServiceUrl>
```
Comment (strongwinderic) [inline src/Docker/docker-compose.dcproj]: old port still present here

**Pull request merged**